File: src/com/maddyhome/idea/vim/command/MappingMode.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2019 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.command;

import java.util.EnumSet;

/**
 * @author vlan
 */
public enum MappingMode {
  /**
   * Indicates this key mapping applies to Normal mode
   */
  NORMAL,
  /**
   * Indicates this key mapping applies to Visual mode
   */
  VISUAL,
  /**
   * Indicates this key mapping applies to Select mode
   */
  SELECT,
  /**
   * Indicates this key mapping applies to Operator Pending mode
   */
  OP_PENDING,
  /**
   * Indicates this key mapping applies to Insert mode
   */
  INSERT,
  /**
   * Indicates this key mapping applies to Command Line mode
   */
  CMD_LINE;

  public static final EnumSet<MappingMode> N = EnumSet.of(NORMAL);
  public static final EnumSet<MappingMode> X = EnumSet.of(VISUAL);
  public static final EnumSet<MappingMode> O = EnumSet.of(OP_PENDING);
  public static final EnumSet<MappingMode> I = EnumSet.of(INSERT);
  public static final EnumSet<MappingMode> C = EnumSet.of(CMD_LINE);
  public static final EnumSet<MappingMode> S = EnumSet.of(SELECT);
  public static final EnumSet<MappingMode> V = EnumSet.of(VISUAL, SELECT);

  public static final EnumSet<MappingMode> NO = EnumSet.of(NORMAL, OP_PENDING);
  public static final EnumSet<MappingMode> XO = EnumSet.of(VISUAL, OP_PENDING);
  public static final EnumSet<MappingMode> NX = EnumSet.of(NORMAL, VISUAL);
  public static final EnumSet<MappingMode> IC = EnumSet.of(INSERT, CMD_LINE);
  public static final EnumSet<MappingMode> NV = EnumSet.of(NORMAL, VISUAL, SELECT);

  public static final EnumSet<MappingMode> NXO = EnumSet.of(NORMAL, VISUAL, OP_PENDING);
  public static final EnumSet<MappingMode> NVO = EnumSet.of(NORMAL, VISUAL, OP_PENDING, SELECT);

  public static final EnumSet<MappingMode> ALL = EnumSet.allOf(MappingMode.class);
}

File: vim-engine/src/main/kotlin/com/maddyhome/idea/vim/command/MappingMode.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.command

import com.maddyhome.idea.vim.ex.ExException
import java.util.*

/**
 * @author vlan
 *
 * COMPATIBILITY-LAYER: Do not move this class to a different package
 */
enum class MappingMode {
  /**
   * Indicates this key mapping applies to Normal mode
   */
  NORMAL,

  /**
   * Indicates this key mapping applies to Visual mode
   */
  VISUAL,

  /**
   * Indicates this key mapping applies to Select mode
   */
  SELECT,

  /**
   * Indicates this key mapping applies to Operator Pending mode
   */
  OP_PENDING,

  /**
   * Indicates this key mapping applies to Insert mode
   */
  INSERT,

  /**
   * Indicates this key mapping applies to Command Line mode
   */
  CMD_LINE,

  ;

  companion object {
    val N: EnumSet<MappingMode> = EnumSet.of(NORMAL)
    val X: EnumSet<MappingMode> = EnumSet.of(VISUAL)
    val O: EnumSet<MappingMode> = EnumSet.of(OP_PENDING)
    val I: EnumSet<MappingMode> = EnumSet.of(INSERT)
    val C: EnumSet<MappingMode> = EnumSet.of(CMD_LINE)
    val S: EnumSet<MappingMode> = EnumSet.of(SELECT)
    val V: EnumSet<MappingMode> = EnumSet.of(VISUAL, SELECT)
    val IC: EnumSet<MappingMode> = EnumSet.of(INSERT, CMD_LINE)
    // This requires the JvmField annotation as it is used (in Java) by an external plugin
    @JvmField val NVO: EnumSet<MappingMode> = EnumSet.of(NORMAL, VISUAL, OP_PENDING, SELECT)

    // TODO: Consider removing/depracting XO, NV and NXO. They're not typical Vim modes that have a map command
    // E.g. `xmap` is for Visual, `vmap` is for Visual and Select. `map` is NVO and `map!` is IC
    // There are no Vim map commands for XO, NV or NXO. If there isn't a Vim API for them, we shouldn't offer one
    // either. It could lead to confusion or incorrect mapping
    // Note that builtin commands can be valid only in NXO (Normal, Visual and Op-pending) but that doesn't mean it's
    // a valid mapping mode.

    // Used externally, by Java
    @JvmField val XO: EnumSet<MappingMode> = EnumSet.of(VISUAL, OP_PENDING)
    // Used externally
    val NV: EnumSet<MappingMode> = EnumSet.of(NORMAL, VISUAL, SELECT)
    // Used externally
    val NXO: EnumSet<MappingMode> = EnumSet.of(NORMAL, VISUAL, OP_PENDING)

    // This method is used only for single modes, not groups of them (V is not supported)
    fun parseModeChar(char: Char): MappingMode {
      return when (char.uppercaseChar()) {
        'N' -> NORMAL
        'X' -> VISUAL
        'S' -> SELECT
        'O' -> OP_PENDING
        'I' -> INSERT
        'C' -> CMD_LINE
        else -> throw ExException("Unexpected mode for char $char")
      }
    }

    fun Set<MappingMode>.toModeString(): String {
      if (this == MappingMode.IC) return "!"
      if (this == MappingMode.NVO) return " "
      if (this == MappingMode.C) return "c"
      if (this == MappingMode.I) return "i"
      //if (modes.equals(MappingMode.L)) return "l";

      // The following modes are concatenated
      val modes = this
      return buildString {
        if (modes.containsAll(MappingMode.N)) append("n")
        if (modes.containsAll(MappingMode.O)) append("o")
        if (modes.containsAll(MappingMode.V)) {
          append("v")
        } else {
          if (modes.containsAll(MappingMode.X)) append("x")
          if (modes.containsAll(MappingMode.S)) append("s")
        }
      }
    }

    fun fromModeString(modes: String): Set<MappingMode> {
      if (modes == "!") return MappingMode.IC
      if (modes == " ") return MappingMode.NVO

      if (modes.length == 2 && modes.contains("i") && modes.contains("c")) {
        return MappingMode.IC
      }
      if (modes.length == 3 && modes.contains("n") && modes.contains("v") && modes.contains("o")) {
        return MappingMode.NVO
      }

      return buildSet<MappingMode> {
        modes.forEach {
          when (it) {
            'c' -> addAll(MappingMode.C)
            'i' -> addAll(MappingMode.I)
            'n' -> addAll(MappingMode.N)
            'o' -> addAll(MappingMode.O)
            'v' -> addAll(MappingMode.V)
            'x' -> addAll(MappingMode.X)
            's' -> addAll(MappingMode.S)
          }
        }
      }
    }
  }
}
